hutch_python/obj_config.py

```diff
@@ -0,0 +1,189 @@
+import logging
+from pathlib import Path
+from typing import Callable, Tuple, Union
+
+import yaml
+from ophyd import Kind
+from pcdsdevices.interface import BaseInterface
+
+from . import utils
+
+logger = logging.getLogger(__name__)
+
+
+def update_objs(
+    obj_ns: utils.HelpfulNamespace,
+    entry: str,
+    attrs: dict,
+    fn: Callable[[BaseInterface, Union[dict, list[str]]], None]
+) -> None:
+    """
+    Helper function for updating object namespaces
+    obj_ns is a HelpfulNamespace, which may include other
+    HelpfulNamespace's.  As a result this must be semi-recursive.
+
+    Parameters
+    ----------
+    obj_ns : HelpfulNamespace
+        Namespace of all objects
+
+    entry : str
+        Entry in obj_conf yaml. Name of device or class of devices to edit.
+
+    attrs : list[str]
+        List of attributes to add to whitelist
+
+    fn : Callable[[BaseInterface, dict], None]
+        A function that applies a change to a singular device
+    """
+    dev = getattr(obj_ns, entry, None)
+    if dev:  # look for explicit device
+        fn(dev, attrs)
+
+        return
+
+    # look for devices with class
+    for dev in obj_ns:
+        if isinstance(dev, utils.HelpfulNamespace):
+            update_objs(dev, entry, attrs, fn)
+        # a poor-man's type check, in an effort to avoid importing
+        # every device type again
+        if entry == type(dev).__name__:
+            fn(dev, attrs)
+
+
+def update_whitelist(dev: BaseInterface, attrs: list[str]) -> None:
+    """
+    Update the tab whitelist for ``dev`` with each field in ``attrs``
+
+    Parameters
+    ----------
+    dev : BaseInterface
+        A single object implementing the tab completion features in
+        pcdsdevices.interface.BaseInterface
+
+    attrs : list[str]
+        list of attributes to add to ``dev`` tablist
+    """
+    for att in attrs:
+        dev._tab.add(att)
+
+
+def update_blacklist(dev: BaseInterface, attrs: list[str]) -> None:
+    """
+    Remove each field in ``attrs`` from tab whitelist for ``dev``
+
+    Parameters
+    ----------
+    dev : BaseInterface
+        A single object implementing the tab completion features in
+        pcdsdevices.interface.BaseInterface
+
+    attrs : list[str]
+        list of attributes to remove from ``dev`` tablist
+    """
+    for att in attrs:
+        try:
+            dev._tab.remove(att)
+        except KeyError:
+            logger.debug(f'key: {att} not in tab completion list')
+
+
+def replace_tablist(dev: BaseInterface, attrs: list[str]) -> None:
+    """
+    Replace tab completion list for ``dev`` with ``attrs``
+
+    Parameters
+    ----------
+    dev : BaseInterface
+        A single object implementing the tab completion features in
+        pcdsdevices.interface.BaseInterface
+
+    attrs : list[str]
+        list of attributes to replace ``dev`` tablist with
+    """
+    for old_att in dir(dev):
+        dev._tab.remove(old_att)
```

Comment:
I don't think this is exhaustive, since the tab completion mechanism lets us use regex as specifiers. It'd be easiest to just call reset I think: https://github.com/pcdshub/pcdsdevices/blob/4c01ba8892215c535e7e6ee2ac5a418f5aea4a07/pcdsdevices/interface.py#L71

Reply:
I'm running into problems using `_tab.reset()`, as the failing CI attests to.  Phenomenologically, `reset` seems to just revert the tab settings to the default state, but looking at the code I'd expect it to clear out all the settings. There's some strange subclassing inheritance stuff going that I don't understand, maybe?

Reply:
I can read, and understand better now.  `_tab.remove` is better than `_tab.reset()`, but still doesn't catch regex examples.